Detectors/Calibration/include/DetectorsCalibration/TimeSlotCalibration.h

```diff
@@ -112,27 +128,67 @@ template <typename Input, typename Container>
 void TimeSlotCalibration<Input, Container>::checkSlotsToFinalize(TFType tf, int maxDelay)
 {
   // Check which slots can be finalized, provided the newly arrived TF is tf
-  // check if some slots are done
-  for (auto slot = mSlots.begin(); slot != mSlots.end(); slot++) {
-    if (maxDelay == 0 || (slot->getTFEnd() + maxDelay) < tf) {
-      if (hasEnoughData(*slot)) {
-        finalizeSlot(*slot); // will be removed after finalization
-      } else if ((slot + 1) != mSlots.end()) {
-        LOG(INFO) << "Merging underpopulated slot " << slot->getTFStart() << " <= TF <= " << slot->getTFEnd()
-                  << " to slot " << (slot + 1)->getTFStart() << " <= TF <= " << (slot + 1)->getTFEnd();
-        (slot + 1)->mergeToPrevious(*slot);
+
+  constexpr uint64_t INFINITE_TF = 0xffffffffffffffff;
+  constexpr int64_t INFINITE_TF_int64 = std::numeric_limits<long>::max() - 1; // this is used to define the end
+                                                                              // of the slot in case it is "std::numeric_limits<long>::max()"
+                                                                              // long (so we need to subtract 1)
+
+  // if we have one slot only which is INFINITE_TF_int64 long, and we are not at the end of run (tf != INFINITE_TF),
+  // we need to check if we got enough statistics, and if so, redefine the slot
+  if (mSlots.size() == 1 && mSlots[0].getTFEnd() == INFINITE_TF_int64) {
+    if (tf >= mUpdateInterval + mLastCheckedTF || tf == INFINITE_TF) {
```

Comment:
not sure this is correct: `mLastCheckedTF` will be updated to the new tf and the next check will be postponed by another mUpdateInterval TFs. Should not you use the lowest checked TF (should be introduced)?

Reply:
Ciao @noferini , I think that Ruben understood correctly: what I want is that the next time the check happens is only after `mUpdateInterval`, from the last check, so it checks every `mUpdateInterval` TFs (more or less, since they are async). I don't understand what one would gain using the lowest checked TF.

Reply:
Exactly, I thought if you consider that after 5 min it is worth to check the stat., then from that moment on you should keep checking. 5 min is an easy case, but if e.g. you put the mUpdateInterval to 1h, then the next check will happen only in 2h.
Anyway, this is a matter of conventions, can be left also like this. But the names are somewhat misleading, I would rename 
`mUpdateInterval` to `mCheckIntervalInfSlot` and `mLastCheckedTFInfSlot`, to indicate that this applies to infinite single slot processing case.

Reply:
Ok, I can change the naming. When I tested, checking at every TF was not making sense, but maybe one could check with a difference frequency if the first check did not pass. This would require a new variable, but it is not an issue. What do you think? E.g. a fraction of the update interval.

Reply:
Yes, I understood, but if your mUpdateInterval is just a guess, it will check at every TF for long. This was my first implementation, and I did not like it.

Reply:
This was a my proposal since I didn't understand what is the gain to check only after a given threshold.
I started assuming that the problem was the overhead introduced by the check. For that I think that the best solution is to dilute such an operation (even if you expect to have enough statistics after 1h you can aslo check every 5 minutes since it is not a problem in terms of CPU).
If there is another reason to have such a threshold we can re-arrange name in a different way. 
If the overhead expected for the check is not a problem we can restore as it was before.

Reply:
The code initially checked at every TF. I changed it because if one expects that anyway it takes some time to reach enough statistics, then he can use this knowledge and tune the code. 
hasEnoughData should take longer than the `if's` above, e.g. for TOF we check the value of ~240K elements of an array. It is not super expensive, but why doing it if you know there is anyway no chance? In addition, setting this value to 1 would check at every TF, so that type of behavior is there.

If you think this does not make sense, I will remove it, but I think it could be beneficial. And: one can go with a solution in between: check ones after mUpdateInterval and if not successful check again with a different frequency. 

Let me know what you prefer, I tried to be flexible, but I will not insist (a check every time frame was bothering me very much in debugging phase, but this is a special case of course).

Reply:
Hi @chiarazampolli 

It does not hurt, since with default value mUpdateInterval to 1 it is equivalent to checking every TF. 
Concerning your use case of hasEnoughData in TOF: checking 240k elements is quite expensive, but why do you need to do this? When you fill the channel you can increment the counter of "completed channels" if it has reached the target stat., and in the hasEnoughData you just check this counter.

Reply:
> since with default value mUpdateInterval to 1 it is equivalent to checking every TF
Yes, this is also what I mentioned. I had already thought about an "intermediate" setting, maybe I will add it. Someone might use it. 
> When you fill the channel you can increment the counter of "completed channels" if it has reached the target stat., and in the hasEnoughData you just check this counter.
Good suggestion, will do it on Monday!

Reply:
Hi @shahor02 , @noferini ,
I introduced the possibility to retry the calibration at a different frequency, if the previous check failed. 
For now, I did not change the `hasEnoughData`, since the current code does not only count the number of entries, but also defines a mean over the channels that have at least 1 entry. So, if I move this part in the `fill`, I will need to do several `if`'s there, which might not be optimal. E.g.:
```
for (int i = data.size(); i--;) {
auto ch = data[i].getTOFChIndex();
...
if (mEntries[ch] == 0) mNValid++;
++mMean;
mEntries[ch] += 1;
}
```
Francesco, the occupancy at TOF in PbPb is how much? I think we usually said 100 useful tracks/collisions. Then, if we want to have 500 tracks per channel for beam data, we need (500 * 160000)/100 events, which, with IR = 30 kHz (average), means ~26 seconds. This is then ~2700 TFs. So, before getting enough statistics, I would do 2700 * 100 times `if (mEntries[ch] == 0) mNValid++;`. Right? 
At the moment, the `hasNoughtData`has some `if`s inside a loop over the number of channels (which is 160K in case of beam data, and 240K in case of cosmics), but we could even remove them as they are not really used. Then we would be left with the loop over the number of channels, which I think is less expensive than moving everything in the `fill` method. What do you think? 
@noferini , if needed, I can explain you what I mean in "virtual" person.

Chiara

Reply:
Hi Chiara, we should discuss about that (I agree with you), how to implement the check in HasEnoughData properly. Now I would focus on the chain in order to have something which we can test on EPNs with cosmics to see where bottlenecks could be.

Reply:
Sure, if also for Ruben it is fine, we can leave it for the next PR (where we will also split the fitting from the accumulation of the data - this I mention for Ruben).